com.hazelcast.jet.core.metrics.MetricsTest.usingContextAsync

_master_ (commit fbdc67d2d01f06b7c8e2dce834bad1b95d725cc7)

Failed on Oracle JDK 8: http://jenkins.hazelcast.com/job/jet-oss-master-oracle-jdk8/126/testReport/junit/com.hazelcast.jet.core.metrics/MetricsTest/usingContextAsync/

Stacktrace:
```
java.lang.AssertionError: expected:<100000> but was:<94671>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.jet.core.metrics.MetricsTest.lambda$usingContextAsync$2(MetricsTest.java:201)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1325)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1427)
	at com.hazelcast.jet.core.metrics.MetricsTest.usingContextAsync(MetricsTest.java:201)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at com.hazelcast.test.DumpBuildInfoOnFailureRule$1.evaluate(DumpBuildInfoOnFailureRule.java:37)
	at com.hazelcast.test.jitter.JitterRule$1.evaluate(JitterRule.java:104)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

Standard output:
```
13:27:36,623 DEBUG || - [JobExecutionService] hz.distracted_bhabha.jet.blocking.thread-1 - [127.0.0.1]:5701 [jet] [4.0-SNAPSHOT] Execution of job '0361-e537-e0c0-0001', execution 0361-e537-e0c1-0001 completed with failure
java.util.concurrent.CompletionException: com.hazelcast.jet.JetException: Exception in ProcessorTasklet{filterUsingServiceAsync#22}: java.lang.AssertionError: outbox has unfinished item after successful completeEdge() or finishSnapshotRestore()
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:593)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalCompleteExceptionally(NonCompletableFuture.java:59)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$ExecutionTracker.taskletDone(TaskletExecutionService.java:421)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:272)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.jet.JetException: Exception in ProcessorTasklet{filterUsingServiceAsync#22}: java.lang.AssertionError: outbox has unfinished item after successful completeEdge() or finishSnapshotRestore()
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:336)
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:309)
	... 1 more
Caused by: java.lang.AssertionError: outbox has unfinished item after successful completeEdge() or finishSnapshotRestore()
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:299)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:208)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:329)
	... 3 more
...
Hiccups measured while running test 'usingContextAsync(com.hazelcast.jet.core.metrics.MetricsTest):'
13:27:35, accumulated pauses: 372 ms, max pause: 252 ms, pauses over 1000 ms: 0
13:27:40, accumulated pauses: 74 ms, max pause: 32 ms, pauses over 1000 ms: 0
13:27:45, accumulated pauses: 120 ms, max pause: 10 ms, pauses over 1000 ms: 0
13:27:50, accumulated pauses: 134 ms, max pause: 10 ms, pauses over 1000 ms: 0
13:27:55, accumulated pauses: 202 ms, max pause: 43 ms, pauses over 1000 ms: 0
13:28:00, accumulated pauses: 128 ms, max pause: 9 ms, pauses over 1000 ms: 0
13:28:05, accumulated pauses: 140 ms, max pause: 10 ms, pauses over 1000 ms: 0
13:28:10, accumulated pauses: 139 ms, max pause: 34 ms, pauses over 1000 ms: 0
13:28:15, accumulated pauses: 104 ms, max pause: 10 ms, pauses over 1000 ms: 0
13:28:20, accumulated pauses: 184 ms, max pause: 27 ms, pauses over 1000 ms: 0
13:28:25, accumulated pauses: 166 ms, max pause: 11 ms, pauses over 1000 ms: 0
13:28:30, accumulated pauses: 160 ms, max pause: 15 ms, pauses over 1000 ms: 0
13:28:35, accumulated pauses: 174 ms, max pause: 10 ms, pauses over 1000 ms: 0
13:28:40, accumulated pauses: 69 ms, max pause: 10 ms, pauses over 1000 ms: 0
13:28:45, accumulated pauses: 62 ms, max pause: 4 ms, pauses over 1000 ms: 0
13:28:50, accumulated pauses: 179 ms, max pause: 30 ms, pauses over 1000 ms: 0
13:28:55, accumulated pauses: 93 ms, max pause: 9 ms, pauses over 1000 ms: 0
13:29:00, accumulated pauses: 150 ms, max pause: 11 ms, pauses over 1000 ms: 0
13:29:05, accumulated pauses: 61 ms, max pause: 11 ms, pauses over 1000 ms: 0
13:29:10, accumulated pauses: 65 ms, max pause: 9 ms, pauses over 1000 ms: 0
13:29:15, accumulated pauses: 97 ms, max pause: 10 ms, pauses over 1000 ms: 0
13:29:20, accumulated pauses: 58 ms, max pause: 6 ms, pauses over 1000 ms: 0
13:29:25, accumulated pauses: 99 ms, max pause: 10 ms, pauses over 1000 ms: 0
13:29:30, accumulated pauses: 51 ms, max pause: 5 ms, pauses over 1000 ms: 0
13:29:35, accumulated pauses: 3 ms, max pause: 0 ms, pauses over 1000 ms: 0
```

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Processor.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Processor.java
@@ -154,10 +154,6 @@ public interface Processor {
      * If the call returns {@code false}, it will be called again before
      * proceeding to call any other method. Default implementation returns
      * {@code true}.
-     * <p>
-     * If this method tried to offer to the outbox and the offer call returned
-     * false, this method must also return false and retry to offer in the
-     * next call.
      */
     default boolean tryProcess() {
         return true;

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ProcessorTasklet.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ProcessorTasklet.java
@@ -255,14 +255,13 @@ public class ProcessorTasklet implements Tasklet {
             case PROCESS_INBOX:
                 progTracker.notDone();
                 if (inbox.isEmpty()) {
-                    if (isSnapshotInbox() || processor.tryProcess()) {
-                        assert !outbox.hasUnfinishedItem() : isSnapshotInbox()
-                                ? "Unfinished item before fillInbox call"
-                                : "Processor.tryProcess() returned true, but there's unfinished item in the outbox";
-                        fillInbox();
-                    } else {
-                        return;
+                    if (!isSnapshotInbox()) {
+                        if (!processor.tryProcess()) {
+                            return;
+                        }
+                        outbox.reset();
                     }
+                    fillInbox();
                 }
                 if (!inbox.isEmpty()) {
                     if (isSnapshotInbox()) {
@@ -296,8 +295,8 @@ public class ProcessorTasklet implements Tasklet {
                 progTracker.notDone();
                 if (isSnapshotInbox()
                         ? processor.finishSnapshotRestore() : processor.completeEdge(currInstream.ordinal())) {
-                    assert !outbox.hasUnfinishedItem() :
-                            "outbox has unfinished item after successful completeEdge() or finishSnapshotRestore()";
+                    assert !outbox.hasUnfinishedItem() || !isSnapshotInbox() :
+                            "outbox has an unfinished item after successful finishSnapshotRestore()";
                     progTracker.madeProgress();
                     state = initialProcessingState();
                 }

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceOrderedP.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceOrderedP.java
@@ -62,7 +62,6 @@ public final class AsyncTransformUsingServiceOrderedP<S, T, R> extends AbstractP
     private Traverser<?> currentTraverser = Traversers.empty();
     private int maxAsyncOps;
     private ResettableSingletonTraverser<Watermark> watermarkTraverser = new ResettableSingletonTraverser<>();
-    private boolean tryProcessSucceeded;
 
     @Probe(name = "numInFlightOps")
     private final AtomicInteger asyncOpsCounterMetric = new AtomicInteger();
@@ -124,13 +123,9 @@ public final class AsyncTransformUsingServiceOrderedP<S, T, R> extends AbstractP
 
     @Override
     public boolean tryProcess() {
-        if (tryProcessSucceeded) {
-            tryFlushQueue();
-        } else {
-            emitFromTraverser(currentTraverser);
-        }
+        tryFlushQueue();
         asyncOpsCounterMetric.lazySet(queue.size());
-        return tryProcessSucceeded = !getOutbox().hasUnfinishedItem();
+        return true;
     }
 
     @Override

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceUnorderedP.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceUnorderedP.java
@@ -167,6 +167,9 @@ public final class AsyncTransformUsingServiceUnorderedP<S, T, K, R> extends Abst
 
     @Override
     public boolean tryProcessWatermark(@Nonnull Watermark watermark) {
+        if (getOutbox().hasUnfinishedItem() && !emitFromTraverser(currentTraverser)) {
+            return false;
+        }
         assert lastEmittedWm <= lastReceivedWm : "lastEmittedWm=" + lastEmittedWm + ", lastReceivedWm=" + lastReceivedWm;
         // Ignore a watermark that is going back. This is possible after restoring from a snapshot
         // taken in at-least-once mode.

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceUnorderedP.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceUnorderedP.java
@@ -85,7 +85,6 @@ public final class AsyncTransformUsingServiceUnorderedP<S, T, K, R> extends Abst
     private final Map<T, Integer> inFlightItems = new IdentityHashMap<>();
     private Traverser<Object> currentTraverser = Traversers.empty();
     private Traverser<Entry> snapshotTraverser;
-    private boolean tryProcessSucceeded;
 
     private Long lastReceivedWm = Long.MIN_VALUE;
     private long lastEmittedWm = Long.MIN_VALUE;
@@ -186,13 +185,9 @@ public final class AsyncTransformUsingServiceUnorderedP<S, T, K, R> extends Abst
 
     @Override
     public boolean tryProcess() {
-        if (tryProcessSucceeded) {
-            tryFlushQueue();
-        } else {
-            // if we're running tryProcess for the second time, emit just the current traverser
-            emitFromTraverser(currentTraverser);
-        }
-        return tryProcessSucceeded = !getOutbox().hasUnfinishedItem();
+        tryFlushQueue();
+        asyncOpsCounterMetric.lazySet(asyncOpsCounter);
+        return true;
     }
 
     @Override

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/metrics/MetricsTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/metrics/MetricsTest.java
@@ -33,8 +33,10 @@ import com.hazelcast.jet.pipeline.Pipeline;
 import com.hazelcast.jet.pipeline.ServiceFactory;
 import com.hazelcast.jet.pipeline.Sinks;
 import com.hazelcast.jet.pipeline.test.TestSources;
+import com.hazelcast.test.HazelcastSerialClassRunner;
 import org.junit.Before;
 import org.junit.Test;
+import org.junit.runner.RunWith;
 
 import java.util.Arrays;
 import java.util.HashSet;
@@ -48,6 +50,7 @@ import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 
+@RunWith(HazelcastSerialClassRunner.class)
 public class MetricsTest extends JetTestSupport {
 
     private static final JobConfig JOB_CONFIG_WITH_METRICS = new JobConfig().setStoreMetricsAfterJobCompletion(true);
@@ -195,7 +198,7 @@ public class MetricsTest extends JetTestSupport {
                             );
                         }
                 )
-                .drainTo(Sinks.logger());
+                .drainTo(Sinks.noop());
 
         Job job = instance.newJob(pipeline, JOB_CONFIG_WITH_METRICS);
         assertTrueEventually(() -> assertEquals(inputSize,

